### ISSUE TITLE
We should not send mutable objects to Akka

### ISSUE THREAD

**Issue by kernixski** [kullervo16/checklist#98]
See http://doc.akka.io/docs/akka/current/general/jmm.html

**Comment by kullervo16** [comment]
Indeed, and we don't... the messages are immutable, as requested by akka (so conform the spec).

But we do send a key (immutable) by which we retrieve the state object (mutable)... but that's ok, since the request is to persist **the most recent state**. Is it possible that we already have changed the state before the message is processed : yes, is that a problem : no (we are not building a versioning system... if we would do that, yes it would be a problem since we would have missed a version... but with the current implementation : perfectly valid, and it saves us the job of creating a deep clone and pass that via the message. When we do change to versioning : yes we will need to create the clone and ship that in the message).

**Comment by kernixski** [comment]
The problem is not about versioning but about the state of the object in memory with multi-threading.  When you work with multiple threads, you have to synchronize the access to the objects.  Or, you can use immutable objects just as asked in the Akka specs.

The solution is to create an immutable object with all fields declared as final and then, indeed, initialized in the constructor.  This is the way to avoid problems with multi-threading.  You can find a lot of posts talking about the Java Memory Barrier on the net.  Or we can discuss about that tomorrow.

If you want an example, if a thread update an object with 2 non final fields, the second thread will maybe have one of the 2 fields non updated even if the thread 2 read the 2 properties after the end of the updates made by the thread one.  The JVM does not guarantee the 2 fields will be updated in the CPU cache used by the thread 2.  That's why Akka ask for immutable objects.
